modules/rostests/apitests/advapi32/Hash.c

```diff
@@ -0,0 +1,298 @@
+
+typedef struct _A_SHA_CTX
+{
+    UCHAR Buffer[64];
+    ULONG State[5];
+    ULONG Count[2];
+} A_SHA_CTX, *PA_SHA_CTX;
+
+#define A_SHA_DIGEST_LEN 20
+
+typedef struct _MD5_CTX
+{
+    ULONG Count[2];
+    ULONG State[4];
+    UCHAR Buffer[64];
+    UCHAR Hash[16];
+} MD5_CTX, *PMD5_CTX;
+
+#define MD5_DIGEST_LEN 16
+
+typedef struct _MD4_CTX
+{
+    ULONG State[4];
+    ULONG Count[2];
+    UCHAR Buffer[64];
+    UCHAR Hash[16];
+} MD4_CTX, *PMD4_CTX;
+
+#define MD4_DIGEST_LEN 16
+
+#ifndef RSA32API
+#define RSA32API __stdcall
+#endif
+
+typedef
+VOID
+RSA32API
+FN_A_SHAInit(
+    _Out_ PA_SHA_CTX Context);
+
+typedef
+VOID
+RSA32API
+FN_A_SHAUpdate(
+    _Inout_ PA_SHA_CTX Context,
+    _In_reads_(BufferSize) PUCHAR Buffer,
+    _In_ ULONG BufferSize);
+
+typedef
+VOID
+RSA32API
+FN_A_SHAFinal(
+    _Inout_ PA_SHA_CTX Context,
+    _Out_ PUCHAR Result);
+
+typedef
+VOID
+RSA32API
+FN_MD5Init(
+    _Out_ PMD5_CTX Context);
+
+typedef
+VOID
+RSA32API
+FN_MD5Update(
+    _Inout_ PMD5_CTX Context,
+    _In_reads_(BufferSize) PUCHAR Buffer,
+    _In_ ULONG BufferSize);
+
+typedef
+VOID
+RSA32API
+FN_MD5Final(
+    _Inout_ PMD5_CTX Context);
+
+typedef
+VOID
+RSA32API
+FN_MD4Init(
+    _Out_ PMD4_CTX Context);
+
+typedef
+VOID
+RSA32API
+FN_MD4Update(
+    _Inout_ PMD4_CTX Context,
+    _In_reads_(BufferSize) PUCHAR Buffer,
+    _In_ ULONG BufferSize);
+
+typedef
+VOID
+RSA32API
+FN_MD4Final(
+    _Inout_ PMD4_CTX Context);
+
+static HMODULE g_hAdvapi32 = NULL;
+static ANSI_STRING g_TestString = RTL_CONSTANT_STRING("ReactOS Hash API Test String");
+
+static ULONG g_ctxSHA1StateInit[] = { 0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0 };
+static UCHAR g_aucSHA1Result[A_SHA_DIGEST_LEN] = {
+    0xEC, 0x05, 0x43, 0xE7, 0xDE, 0x8A, 0xEE, 0xFF,
+    0xAD, 0x72, 0x2B, 0x9D, 0x55, 0x4F, 0xCA, 0x6A,
+    0x8D, 0x81, 0xF1, 0xC7
+};
+
+static ULONG g_aulMD5Or4StateInit[] = { 0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476 };
+static UCHAR g_aucMD5Result[MD5_DIGEST_LEN] = {
+    0x3D, 0xE8, 0x23, 0x8B, 0x9D, 0xE0, 0xCE, 0x48,
+    0xB1, 0x1B, 0xDD, 0xD9, 0xC6, 0x86, 0xB2, 0xDE
+};
+static UCHAR g_aucMD4Result[MD4_DIGEST_LEN] = {
+    0xE0, 0xE8, 0x50, 0x8A, 0x4D, 0x11, 0x02, 0xA6,
+    0x6A, 0xF0, 0xA7, 0xAB, 0xD8, 0xC4, 0x40, 0xED
+};
+
+static void Test_SHA1(void)
+{
+    FN_A_SHAInit* pfnA_SHAInit;
+    FN_A_SHAUpdate* pfnA_SHAUpdate;
+    FN_A_SHAFinal* pfnA_SHAFinal;
+    DECLSPEC_ALIGN(4) A_SHA_CTX ctx;
+    UCHAR Result[A_SHA_DIGEST_LEN];
+
+    /* Load functions */
+    pfnA_SHAInit = (FN_A_SHAInit*)GetProcAddress(g_hAdvapi32, "A_SHAInit");
+    pfnA_SHAUpdate = (FN_A_SHAUpdate*)GetProcAddress(g_hAdvapi32, "A_SHAUpdate");
+    pfnA_SHAFinal = (FN_A_SHAFinal*)GetProcAddress(g_hAdvapi32, "A_SHAFinal");
+
+    if (!pfnA_SHAInit || !pfnA_SHAUpdate || !pfnA_SHAFinal)
+    {
+        skip("advapi32.dll!A_SHAXXX not found\n");
+        return;
+    }
+
+    /* Test A_SHAInit */
+    pfnA_SHAInit(&ctx);
+    ok(RtlCompareMemory(ctx.State,
```

Comment:
I hope the ok() macro works correctly!
Because, as noticed by Timo for another ok_... one, putting a function call directly inside it would evaluate the argument more than once.
(This is why it might be safer to put the function call outside of it.)

Reply:
Oh, it's not right indeed, I'm thoughtless here, and the test will report correct result in Windows by accident.

I put the function call outside now, here and other places.

The result is the same on Windows:
![20240402042219](https://github.com/reactos/reactos/assets/55521781/2261f97e-069e-4cab-80d6-cc68ff306542)